### DISCUSSION TITLE
Grafik Cpp Bibliothek Diskussion

### DISCUSSION THREAD

**Discussion by Postea** [TRNO-ENTERPRISE/treno-game#6] [Foundations]
## Übersicht: 
Welche cpp lib zur einfachen Visualisierung von treno wollen wir verwenden ?

**Comment by cancoel** [comment]
Hier habe ich eine Liste von Libs gefunden: https://github.com/raizam/gamedev_libraries

**Comment by Postea** [comment]
# Simple DirectMedia Layer (SDL)

## Allgemein
Eine sehr einfache Grafik und Window-Management Bibliothek ähnlich von Swing oder SWT. Es lässt sich sehr einfach verwenden und ist sehr schlank. Definitiv mächtig genug für unsere ersten Meilensteine.

## Relevanz
Gibt uns eine sehr einfache quick-and-dirty Visualisierung. Es benötigt so gut wie keine Einarbeitung und hat den Vorteil, dass wir genau kontrollieren was im Hintergrund passiert.

## Beispiele
```
#include <SDL2/SDL.h>
#include <SDL2/SDL_image.h>
#include <SDL2/SDL_timer.h>
 
int main(int argc, char *argv[])
{
 
    // returns zero on success else non-zero
    if (SDL_Init(SDL_INIT_EVERYTHING) != 0) {
        printf("error initializing SDL: %s\n", SDL_GetError());
    }
    SDL_Window* win = SDL_CreateWindow("GAME",
                                       SDL_WINDOWPOS_CENTERED,
                                       SDL_WINDOWPOS_CENTERED,
                                       1000, 1000, 0);
    while (1)
        ;
 
    return 0;
}
```

Einfaches Beispiel, wie man ein leeres Fenster angezeigt bekommt.

## Links
https://www.geeksforgeeks.org/sdl-library-in-c-c-with-examples/
https://www.libsdl.org

**Comment by cancoel** [comment]
# Magnum Engine (Middleware)

## Allgemein
- Geeignet für Hobby Game Developer
- Open Source
- plattformunabhängig
- gute Dokumentation
- enthält Bibliothek für Vektormathematik und C++11 wrapper für OpenGL API
- Debugging und Testautomation möglich
- UI Bibliothek enthalten
- erweiterbar durch Plugins

## Relevanz
- Bibliothekenwahl ist frei, man kann auch eigene Bibliotheken einbinden
- Durch die ganzen Bibs, die mitgeliefert werden, lässt sich Magnum als Engine verwenden - muss aber nicht. Man benutzt, was man möchte

## Beispiele
- https://magnum.graphics/showcase/

## Links
- https://magnum.graphics
